File: javascript/sorting.js
/**
 * Quicksort algorithm implementation.
 */
function quicksort(arr) {
    if (arr.length <= 1) {
        return arr;
    }
    
    const pivot = arr[Math.floor(arr.length / 2)];
    const left = arr.filter(x => x < pivot);
    const middle = arr.filter(x => x === pivot);
    const right = arr.filter(x => x > pivot);
    
    return [...quicksort(left), ...middle, ...quicksort(right)];
}

/**
 * Run sorting benchmark with 100,000 random integers.
 */
function runBenchmark() {
    // Generate 100,000 random integers
    const arr = Array.from({ length: 100000 }, () => Math.floor(Math.random() * 1000000));
    
    // Measure sorting time
    const startTime = performance.now();
    const sortedArr = quicksort(arr);
    const endTime = performance.now();
    
    const executionTime = endTime - startTime;
    
    // Verify sorting correctness
    const isSorted = sortedArr.every((val, i, arr) => i === 0 || arr[i - 1] <= val);
    
    return {
        testName: 'Sorting (Quicksort)',
        executionTimeMs: executionTime,
        size: arr.length,
        correct: isSorted
    };
}

if (require.main === module) {
    const result = runBenchmark();
    console.log(`Test: ${result.testName}`);
    console.log(`Array size: ${result.size}`);
    console.log(`Execution time: ${result.executionTimeMs.toFixed(2)} ms`);
    console.log(`Correctly sorted: ${result.correct}`);
}

module.exports = { runBenchmark };

File: cpp/sorting.cpp
#include <iostream>
#include <vector>
#include <random>
#include <chrono>
#include <algorithm>

// Quicksort algorithm implementation
std::vector<int> quicksort(std::vector<int> arr) {
    if (arr.size() <= 1) {
        return arr;
    }

    int pivot = arr[arr.size() / 2];
    std::vector<int> left, middle, right;

    for (int x : arr) {
        if (x < pivot) {
            left.push_back(x);
        } else if (x == pivot) {
            middle.push_back(x);
        } else {
            right.push_back(x);
        }
    }

    std::vector<int> sorted_left = quicksort(left);
    std::vector<int> sorted_right = quicksort(right);

    std::vector<int> result = sorted_left;
    result.insert(result.end(), middle.begin(), middle.end());
    result.insert(result.end(), sorted_right.begin(), sorted_right.end());

    return result;
}

// Verify if array is sorted
bool is_sorted(const std::vector<int>& arr) {
    for (size_t i = 0; i < arr.size() - 1; i++) {
        if (arr[i] > arr[i + 1]) {
            return false;
        }
    }
    return true;
}

// Run sorting benchmark with 100,000 random integers
void run_benchmark() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, 999999);

    // Generate 100,000 random integers
    std::vector<int> arr(100000);
    for (int& val : arr) {
        val = dis(gen);
    }

    // Measure sorting time
    auto start_time = std::chrono::high_resolution_clock::now();
    std::vector<int> sorted_arr = quicksort(arr);
    auto end_time = std::chrono::high_resolution_clock::now();
    auto execution_time = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time).count();

    // Verify sorting correctness
    bool correct = is_sorted(sorted_arr);

    std::cout << "Test: Sorting (Quicksort)" << std::endl;
    std::cout << "Array size: " << arr.size() << std::endl;
    std::cout << "Execution time: " << execution_time << " ms" << std::endl;
    std::cout << "Correctly sorted: " << (correct ? "true" : "false") << std::endl;
}

int main() {
    run_benchmark();
    return 0;
}
